grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/InIsaFragment.java

```diff
@@ -54,22 +54,30 @@
     @Override
     public abstract Var end();
 
+    abstract boolean mayHaveEdgeInstances();
+
     @Override
     public GraphTraversal<Vertex, ? extends Element> applyTraversalInner(
             GraphTraversal<Vertex, ? extends Element> traversal, GraknTx graph, Collection<Var> vars) {
 
-        GraphTraversal<Vertex, Vertex> isImplicitRelationType =
-                __.<Vertex>hasLabel(RELATIONSHIP_TYPE.name()).has(IS_IMPLICIT.name(), true);
+        GraphTraversal<Vertex, Vertex> vertexTraversal = Fragments.isVertex(traversal);
+
+        if (mayHaveEdgeInstances()) {
+            GraphTraversal<Vertex, Vertex> isImplicitRelationType =
```

Comment:
DO you have the RelationType object when executing this traversal? Because if so we have another DB read which can be avoided. This is all cached in the core API.

Reply:
No, I don't. I would have to build it mid-traversal - which is possible but it would involve the lambda step at Marko would be sad :(